omnibus/resources/agent/msi/cal/CustomAction.cpp

```diff
@@ -0,0 +1,285 @@
+#include "stdafx.h"
+
+static std::wstring secretUserUsername(L"secretuser");
+static std::wstring secretUserDescription(L"DataDog Secret User");
+static std::wstring datadog_path = L"Datadog\\Datadog Agent";
+static std::wstring datadog_key_root = L"SOFTWARE\\" + datadog_path;
+static std::wstring datadog_key_secret_key = L"secrets";
+static std::wstring datadog_key_secrets = L"SOFTWARE\\" + datadog_path + L"\\" + datadog_key_secret_key;
+static std::wstring datadog_acl_key_secrets = L"MACHINE\\" + datadog_key_secrets;
+
+
+int CreateUser(std::wstring& name, std::wstring& comment, bool writePassToReg = false);
+DWORD DeleteUser(std::wstring& name);
+DWORD DeleteSecretsRegKey();
+
+#ifdef CA_CMD_TEST
+#define LOGMSG_STANDARD 0
+void WcaLog(int type, const char * fmt...)
+{
+	va_list args;
+	va_start(args, fmt);
+	vprintf(fmt, args);
+	va_end(args);
+	printf("\n");
+}
+#else
+extern "C" UINT __stdcall AddDatadogSecretUser(
+	MSIHANDLE hInstall
+	)
+{
+	HRESULT hr = S_OK;
+	UINT er = ERROR_SUCCESS;
+
+	// that's helpful.  WcaInitialize Log header silently limited to 32 chars 
+	hr = WcaInitialize(hInstall, "CA: AddDatadogSecretUser");
+	ExitOnFailure(hr, "Failed to initialize");
+
+	WcaLog(LOGMSG_STANDARD, "Initialized.");
+
+	er = CreateUser(secretUserUsername, secretUserDescription, true);
+	if (0 != er) {
+		hr = -1;
+	}
+
+LExit:
+	er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;
+	return WcaFinalize(er);
+}
+
+extern "C" UINT __stdcall RemoveDatadogSecretUser(MSIHANDLE hInstall) {
+	HRESULT hr = S_OK;
+	UINT er = ERROR_SUCCESS;
+
+	// that's helpful.  WcaInitialize Log header silently limited to 32 chars
+	hr = WcaInitialize(hInstall, "CA: RemoveDatadogSecretUser");
+	ExitOnFailure(hr, "Failed to initialize");
+
+	WcaLog(LOGMSG_STANDARD, "Initialized.");
+
+	er = DeleteUser(secretUserUsername);
+	if (0 != er) {
+		hr = -1;
+		goto LExit;
+	} 
+	er = DeleteSecretsRegKey();
+	if (0 != er) {
+		hr = -1;
+		goto LExit;
+	}
+
+
+LExit:
+	er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;
+	return WcaFinalize(er);
+
+}
+
+
+
+// DllMain - Initialize and cleanup WiX custom action utils.
+extern "C" BOOL WINAPI DllMain(
+	__in HINSTANCE hInst,
+	__in ULONG ulReason,
+	__in LPVOID
+	)
+{
+	switch(ulReason)
+	{
+	case DLL_PROCESS_ATTACH:
+		WcaGlobalInitialize(hInst);
+		// initialize random number generator
+		srand(GetTickCount());
+		break;
+
+	case DLL_PROCESS_DETACH:
+		WcaGlobalFinalize();
+		break;
+	}
+
+	return TRUE;
+}
+#endif // CA_CMD_TEST
+#define MIN_PASS_LEN 12
+#define MAX_PASS_LEN 18
+
+wchar_t * GeneratePassword() {
+	static wchar_t passbuf[MAX_PASS_LEN + 1];
```

Comment:
So, it needs to be static because otherwise the buffer will be destroyed on return, and then the caller won't be able to use it any more.

We could return a `std::string` (but it would still have to be static), and then it's harder to clear the underlying buffer (which I do later).

Reply:
OK